src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,346 @@
+    });
+
+    const probeResults = await Promise.all(probePromises);
+    const successProbeResults = probeResults.filter((probe) => probe.isSuccess);
+
+    if (!successProbeResults.length) {
+      setCurrentProvider(chainId, getFallbackRpcUrl(chainId));
+
+      return;
+    }
+
+    const probeResultsByBlockNumber = orderBy(successProbeResults, ["blockNumber"], ["desc"]);
+    let bestBlockNumberProbe = probeResultsByBlockNumber[0];
+
+    const probeStats = probeResultsByBlockNumber.map((probe, i, arr) => {
+      let isValid = probe.isSuccess;
+
+      const bestBlockNumber = bestBlockNumberProbe.blockNumber;
+      const currProbeBlockNumber = probe.blockNumber;
+      const nextProbeBlockNumber = arr[i + 1]?.blockNumber;
+
+      // Rare case when RPC returned a block number from the far future
+      if (
+        i === 0 &&
+        currProbeBlockNumber &&
+        nextProbeBlockNumber &&
+        currProbeBlockNumber - nextProbeBlockNumber > BLOCK_FROM_FUTURE_THRESHOLD
+      ) {
+        bestBlockNumberProbe = arr[i + 1];
+        isValid = false;
+      }
+
+      // If the block number is lagging behind the best one
+      if (bestBlockNumber && currProbeBlockNumber && bestBlockNumber - currProbeBlockNumber > BLOCK_LAGGING_THRESHOLD) {
+        isValid = false;
+      }
+
+      return {
+        ...probe,
+        isValid,
+      };
+    });
+
+    const bestResponseTimeValidProbe = minBy(
+      probeStats.filter((probe) => probe.isValid),
+      "responseTime"
+    );
+
+    if (bestResponseTimeValidProbe) {
+      setCurrentProvider(chainId, bestResponseTimeValidProbe.url);
+    }
+
+    if (isDebugMode()) {
+      // eslint-disable-next-line no-console
+      console.table(
+        orderBy(
+          probeStats.map((probe) => ({
+            url: probe.url,
+            isSelected: probe.url === bestResponseTimeValidProbe?.url ? "✅" : "",
+            isValid: probe.isValid ? "✅" : "❌",
+            responseTime: probe.responseTime,
+            blockNumber: probe.blockNumber,
+          })),
+          ["responseTime"],
+          ["asc"]
+        )
+      );
+    }
+  });
+}
+
+function setCurrentProvider(chainId: number, newProviderUrl: string) {
+  if (!trackerState) {
+    throw new Error("RPC tracker state is not initialized");
+  }
+
+  trackerState[chainId].currentBestProviderUrl = newProviderUrl;
+
+  const storageKey = JSON.stringify(getRpcProviderKey(chainId));
+
+  localStorage.setItem(
+    storageKey,
+    JSON.stringify({
+      rpcUrl: newProviderUrl,
+      timestamp: Date.now(),
+    })
+  );
+}
+
+function isDebugMode() {
+  return localStorage.getItem(JSON.stringify(SHOW_DEBUG_VALUES_KEY)) === "true";
+}
+
+async function probeRpc(chainId: number, provider: Provider, providerUrl: string): Promise<ProbeData> {
+  const controller = new AbortController();
+
+  return await Promise.race([
+    sleep(PROBE_FAIL_TIMEOUT).then(() => {
+      controller.abort();
+      throw new Error("Probe timeout");
```

Comment:
it's not a big issue, but this error will always be thrown, because timeout is not cleared

i think it's better to use setTimeout and clear it after the successful RPC request. at first glance there is no need to use Promise.race

Reply:
I think the benefit of the `race` expression is that we can leave this error as it is, and don't care about timers as the other promise wins the race. As for me it makes code more readable. Correct me if i'm wrong

Reply:
in this specific case it looks a bit redundant to me 🤔 because on the timeout the ongoing fetch will also be rejected because of controller.abort(), so both promises will fail in fact (although the first one would be sleep timeout)

Reply:
> both promises will fail in fact

if promise is fulfilled once it can't be rejected. same with request - if it's already returned response - we don't care if smth is happened with it later.

I'm ok with this approach, we use it in several critical places and it works 
https://github.com/search?q=repo%3Agmx-io%2Fgmx-interface%20promise.race&type=code